My Project/clsTranslateWinForms.vb

```diff
@@ -1 +1,638 @@
+        '    Runtime  component name: sdgPlots_tbpPlotsOptions_tbpXAxis_ucrXAxis_grpAxisTitle_ucrPnlAxisTitle_rdoSpecifyTitle
+
+        'split the full component name into 2 parts: parent names & child name
+        '  e.g 'sdgPlots_tbpPlotsOptions_tbpXAxis_ucrXAxis_grpAxisTitle' & '_rdoSpecifyTitle'
+        Dim iTmpIndex As Integer = strComponentName.LastIndexOf("_")
+        Dim strParentNames As String = strComponentName.Substring(0, iTmpIndex)
+        Dim strControlName As String = strComponentName.Substring(iTmpIndex)
+
+        'if the dictionary contains a single key that matches <parents>_<other controls>_<child>,
+        '  then return the component associated with that key
+        Dim lstComponents As List(Of Component) =
+            dctComponents.Where(Function(x) Regex.IsMatch(x.Key, strParentNames & "\w+_\w+" & strControlName)) _
+            .Select(Function(x) x.Value).ToList()
+        If lstComponents.Count = 1 Then
+            clsComponent = lstComponents(0)
+        End If
+
+        Return clsComponent
+    End Function
+
+    '''--------------------------------------------------------------------------------------------
+    ''' <summary>   
+    '''    Returns <paramref name="strText"/> translated into <paramref name="strLanguage"/>. 
+    '''    <para>
+    '''    Translations can be bi-directional (e.g. from English to French or from French to English).
+    '''    If <paramref name="strText"/> is already in the current language, or if no translation 
+    '''    can be found, then returns <paramref name="strText"/>.         
+    '''    </para></summary>
+    '''
+    ''' <param name="strText">          The text to translate. </param>
+    ''' <param name="strLanguage">      The language code to translate to (e.g. 'fr' for French).
+    '''                                 </param>
+    ''' <param name="clsConnection">    An open connection to the SQLite translation database. </param>
+    '''
+    ''' <returns>   <paramref name="strText"/> translated into <paramref name="strLanguage"/>. </returns>
+    '''--------------------------------------------------------------------------------------------
+    Private Shared Function GetDynamicTranslation(strText As String, strLanguage As String, clsConnection As SQLiteConnection) As String
+        If String.IsNullOrEmpty(strText) Then
+            Return ""
+        End If
+
+        Using clsCommand As New SQLiteCommand(clsConnection)
+
+            'in the translation text, convert any single quotes to make them suitable for the SQL command
+            strText = strText.Replace("'", "''")
+
+            'get all translations for the specified form and language
+            'Note: The second `SELECT` is needed because we may sometimes need to translate  
+            '      translated text back to the original text (e.g. from French to English when 
+            '      the dialog language toggle button is clicked).
```

Comment:
@lloyddewit this toggling feature may not always be needed by the product. So I'm not sure if we should implement it at this level? I would suggest the toggling effect be implemented at the product level, its simpler and neater that way(even though I'm yet to go through R-Instat toggling functionality).
For instance CLIMSOFT doesn't have the toggle button.

Reply:
That's a very good point.
I suggest that we:

- Keep the check for multiline text,
- Keep the check for text that doesn't contain any letters
- Move the other checks to the product level

For 'doNotTranslate' we have the ignore file. Should we also have a similar txt file that specifies the patterns to translate dynamically? There are also other possible approaches. 
What would work best for Climsoft?
We can discuss.

Reply:
I agree. We can discuss.

Reply:
Sorry, my comment was intended for your suggestion about line 625 above. I just copied the text over.

Reply:
Here is my comment about the second Select statement:
I don't like this second select statement either. It is confusing and is only required for toggling. I also suspect that it is not water-tight. I need to look for special cases where it could potentially fail.
I think all the database selects should be in TranslateWinForms. I'm not sure if we can cleanly move the functionality to R-Instat.
We can discuss.

Reply:
I addressed this comment in issue #5.